sqlglot/optimizer/annotate_types.py

```diff
@@ -531,3 +533,18 @@ def _annotate_bracket(self, expression: exp.Bracket) -> exp.Bracket:
             self._set_type(expression, exp.DataType.Type.UNKNOWN)
 
         return expression
+
+    def _annotate_typeddiv(self, expression: exp.TypedDiv) -> exp.TypedDiv:
+        self._annotate_args(expression)
+
+        l, r = expression.left, expression.right
+        l_is_int = l.type and l.type.this in exp.DataType.INTEGER_TYPES
+        r_is_int = r.type and r.type.this in exp.DataType.INTEGER_TYPES
+
+        if l_is_int and r_is_int:
+            datatype = exp.DataType.Type.BIGINT
+        else:
+            datatype = exp.DataType.Type.DOUBLE
```

Comment:
i'm not sure if this is always correct - e.g. what if either operand is a `DECIMAL` with specific precision etc?

Reply:
I think the coercion rules for division are tricky.... hm

Reply:
I'll switch this to _maybe_coerce, so at least it maintain existing behavior of Div